Boundary data improvements

moxcodes opened the pull request:
## Proposed changes

Makes some improvements to the boundary data system, moving away from taking the extraction radius via the filenames and reducing the incidence of an error that is confusing for users.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

moxcodes committed:
Remove error for high values of H5LookaheadTimes

moxcodes committed:
Permit ExtractionRadius from yaml in Cce

- Previously, standalone Cce required the extraction radius to be specified in
  the filename

nilsvu commented on src/Evolution/Systems/Cce/OptionTags.hpp:
```diff
@@ -204,8 +209,13 @@ struct ExtractionRadius : db::SimpleTag {
+  static double create_from_options(
+      const std::optional<double> extraction_radius) noexcept {
+    ASSERT(
+        static_cast<bool>(extraction_radius),
```
(just a suggestion) I usually prefer the more expressive `extraction_radius.has_value()`

nilsdeppe replied:
we predominantly use the cast notation

kidder replied:
Didn't the cast notation arise because `boost::optional` doesn't have the equivalent of `has_value` and clang-tidy complained about the implicit cast?  I find `has_value` much easier to read and slightly more type-safe (e.g. if someone converts `std::optional<T>` to `T` they may miss the cast...).  Unless there is a compelling reason not to, I like the suggestion of using `has_value`

nilsdeppe replied:
I wasn't trying to suggest that we should use one over the other. I personally prefer keeping consistency throughout the `*::optional` (we should replace boost optional anyway) but will not object to either usage

kidder replied:
I agree we should be consistent, and replacing boost optional is beyond this PR

nilsvu commented on src/Evolution/Systems/Cce/WorldtubeDataManager.cpp:
```diff
@@ -29,28 +29,37 @@ MetricWorldtubeDataManager::MetricWorldtubeDataManager(
     std::unique_ptr<intrp::SpanInterpolator> interpolator,
-    const bool fix_spec_normalization) noexcept
+    const bool fix_spec_normalization,
+    const std::optional<double> extraction_radius) noexcept
```
remove `const` and `std::move` the `extraction_radius` (or is that a trivial operation for a `std::optional<double>`?)

nilsdeppe replied:
it's trivial as far as I'm aware. std::optional stores a bool and the type

nilsvu commented on src/Evolution/Systems/Cce/WorldtubeDataManager.cpp:
```diff
@@ -197,7 +206,10 @@ bool MetricWorldtubeDataManager::populate_hypersurface_boundary_data(
         get<::Tags::dt<Tags::detail::Lapse>>(interpolated_coefficients_),
         get<Tags::detail::Dr<Tags::detail::Lapse>>(interpolated_coefficients_),
-        buffer_updater_->get_extraction_radius(), l_max_);
+        static_cast<bool>(extraction_radius_)
+            ? *extraction_radius_
+            : buffer_updater_->get_extraction_radius(),
```
```suggestion
        extraction_radius_.value_or(buffer_updater_->get_extraction_radius()),
```
(you have a few more occurrences in this PR)

nilsvu commented on src/Evolution/Systems/Cce/WorldtubeDataManager.cpp:
```diff
@@ -29,28 +29,37 @@ MetricWorldtubeDataManager::MetricWorldtubeDataManager(
+                   buffer_depth_)) {
+    buffer_depth_ =
+        buffer_updater_->get_time_buffer().size() -
+        2 * interpolator_->required_number_of_points_before_and_after();
```
I don't quite understand why you can make this change, can you explain (and perhaps add a comment and/or docs)? It looks like the constructed object may end up with a different `buffer_depth_` than is passed to the constructor, which seems unintuitive to me.

moxcodes replied:
Yea, that's intentional (I'll add a code comment). The point is that the buffer depth must not be so large that the worldtube data cannot fill the buffer, so if the input file requests the buffer be larger than the input file size, this adjusts it to be a buffer of length equal to the input file.

nilsvu commented on src/Evolution/Systems/Cce/WorldtubeDataManager.cpp:
```diff
@@ -29,28 +29,37 @@ MetricWorldtubeDataManager::MetricWorldtubeDataManager(
     std::unique_ptr<intrp::SpanInterpolator> interpolator,
-    const bool fix_spec_normalization) noexcept
+    const bool fix_spec_normalization,
+    const std::optional<double> extraction_radius) noexcept
```
Add `#include <optional>`

nilsvu commented on src/Evolution/Systems/Cce/WorldtubeDataManager.hpp:
```diff
@@ -91,7 +92,8 @@ class MetricWorldtubeDataManager : public WorldtubeDataManager {
       std::unique_ptr<intrp::SpanInterpolator> interpolator,
-      bool fix_spec_normalization) noexcept;
+      bool fix_spec_normalization,
+      std::optional<double> extraction_radius = std::nullopt) noexcept;
```
Is this default necessary? Also, please add a few words about the new argument to the docs.

nilsvu commented on src/Evolution/Systems/Cce/OptionTags.hpp:
```diff
@@ -204,8 +209,13 @@ struct ExtractionRadius : db::SimpleTag {
+    ASSERT(
+        static_cast<bool>(extraction_radius),
+        "Cannot infer an `\"Auto\"` value of ExtractionRadius in this context "
+        "-- please specify an explicit value in the input file.");
```
I think you can avoid this situation by switching the `OptionTags::ExtractionRadius` at compile-time. If I understand correctly, you will either create the `Tags::ExtractionRadius` or the `Tags::H5WorldtubeBoundaryDataManager` from options, never both (since if you created both from options you could never set the extraction radius to "Auto"). So you can add a boolean template parameter to `OptionTags::ExtractionRadius` that has the effect of toggling its type between `double` and `Options::Auto<double>` (and changes the help string), or just create a separate `OptionTags::ExtractionRadiusAuto`. Then you can use the non-auto version here in `Tags::ExtractionRadius` and the auto-version in `Tags::H5WorldtubeBoundaryDataManager`.

moxcodes commented:
I've posted a fixup for the first commit.
While looking over the changes for the second commit, I realized that my initial attempt at this PR would still have errored when the filename was not of the right form to communicate the extraction radius. I've modified the PR so that now the extraction radius is processed entirely in the buffer updater instead, which fixes the problem and is a bit more elegant.
I've squashed those changes because the second commit is now rather different, so should just be reviewed again, sorry for causing a repeat review.
Cheers!

nilsvu commented on src/Evolution/Systems/Cce/WorldtubeDataManager.cpp:
```diff
@@ -47,6 +47,11 @@ MetricWorldtubeDataManager::MetricWorldtubeDataManager(
         "points");
   }
+  // This will actually change the buffer depth in the case where the buffer
+  // depth passed to the constructor is to large for the worldtube file size. In
```
typo: to -> too (same in comment below)

nilsvu commented on src/Evolution/Systems/Cce/WorldtubeBufferUpdater.cpp:
```diff
@@ -108,7 +109,7 @@ MetricWorldtubeH5BufferUpdater::MetricWorldtubeH5BufferUpdater(
       cce_data_filename.substr(r_pos + 1, dot_pos - r_pos - 1);
   try {
-    extraction_radius_ = stod(text_radius);
+    extraction_radius_ = extraction_radius.value_or(stod(text_radius));
```
I think this piece of code will incorrectly try to parse the text and might fail, even if the `extraction_radius` has a value. I may have been wrong to suggest `value_or` here because it doesn't evaluate its argument lazily, as a ternary operator would. Whichever way you choose to resolve this issue, please add a test for this case.

nilsvu commented on tests/InputFiles/Cce/CharacteristicExtract.yaml:
```diff
@@ -27,6 +27,7 @@ Cce:
     BarycentricRationalSpanInterpolator:
       MinOrder: 10
       MaxOrder: 10
+  ExtractionRadius: 257.0
```
Do I understand correctly that this can/should be set to `Auto` because the radius can be parsed from the file?

moxcodes replied:
It is true that it can be set to `Auto`, but I want to move away from the interface where numeric parameters are obtained from filenames (that was a SpEC thing...), so I'd prefer to show the example interface the way I think the code should be used most often. Supporting `Auto` is more to ease the transition for those that are still typically importing SpEC data.
Is that reasonable? Would you prefer the example have a filename that can't be parsed if it shows a numeric `ExtractionRadius`?

nilsvu replied:
I think it's fine as-is, thanks for clarifying :)

moxcodes commented:
fixups (`fixup 2:..`) posted for the last couple of comments.
Cheers!

nilsdeppe commented on src/Evolution/Systems/Cce/WorldtubeBufferUpdater.cpp:
```diff
@@ -295,13 +299,14 @@ BondiWorldtubeH5BufferUpdater::BondiWorldtubeH5BufferUpdater(
-    extraction_radius_ = stod(text_radius);
+    extraction_radius_ = static_cast<bool>(extraction_radius)
+                             ? *extraction_radius
+                             : stod(text_radius);
```
`std::stod`?

nilsdeppe replied:
you seem to have replaced one stod (on line 114) with std::stod but not the other :P

moxcodes replied:
d'oh. fixed.

nilsdeppe commented on src/Evolution/Systems/Cce/WorldtubeDataManager.cpp:
```diff
@@ -37,20 +37,32 @@ MetricWorldtubeDataManager::MetricWorldtubeDataManager(
+  // This will actually change the buffer depth in the case where the buffer
+  // depth passed to the constructor is too large for the worldtube file size.
+  // In that case, the worldtube data wouldn't be able to fill the buffer, so
+  // here we shrink the buffer depth down to be no larger than the length of the
```
Yay! I've hit this error message and was :disappointed:

nilsdeppe reviewed (commented):
Just one minor suggestion. Looks fine to me :+1:

@nilsleiffischer could you look at the fixups please?

moxcodes commented:
Last change squashed into the previous fixups.
Cheeers!

nilsvu reviewed (commented):
LGTM please squash.

moxcodes commented:
Fixups squashed in.
Cheers!

kidder commented:
@moxcodes your test timed out 3 different builds, might need to increase the time limit or make the test run faster

moxcodes committed:
Reduce resolution to make Worldtube test faster

moxcodes commented:
Added a commit that scales down the resolution, it seems to make the test run sufficiently fast now.
I've also run the test for 1000 iterations locally to verify that the decreased resolution doesn't cause occasional failures.
Cheers!

nilsdeppe commented:
@nilsleiffischer @kidder could you please take a look at this?
